File: backend/src/utils/PasswordService.js
const crypto = require('crypto');

/**
 * Servicio de gestión segura de contraseñas en backend
 * Usa SHA-256 iterado (compatible con lo que hace el cliente Flutter)
 */
class PasswordService {
  // Número de iteraciones para PBKDF2-like
  static HASH_ITERATIONS = 100000;

  /**
   * Generar salt aleatorio (16 bytes = 32 caracteres hex)
   */
  static generateSalt() {
    const timestamp = Date.now().toString();
    return crypto.createHash('sha256').update(timestamp).digest('hex').substring(0, 32);
  }

  /**
   * Hash seguro de contraseña con salt
   * Retorna: "salt$hash" para guardar en BD
   */
  static hashPassword(password) {
    if (!password || password.length === 0) {
      throw new Error('Contraseña no puede estar vacía');
    }

    const salt = this.generateSalt();
    const hash = this._pbkdf2Like(password, salt);
    return `${salt}$${hash}`;
  }

  /**
   * PBKDF2 simulado usando SHA-256 iterado
   * Compatible con la implementación de Flutter
   */
  static _pbkdf2Like(password, salt) {
    let hash = password;

    for (let i = 0; i < this.HASH_ITERATIONS; i++) {
      hash = crypto
        .createHash('sha256')
        .update(`${hash}${salt}`)
        .digest('hex')
        .substring(0, 64);
    }

    return hash;
  }

  /**
   * Verificar contraseña contra hash almacenado
   * Retorna: true si contraseña es correcta, false en caso contrario
   */
  static verifyPassword(password, storedHash) {
    if (!storedHash || !storedHash.includes('$')) {
      return false;
    }

    try {
      const parts = storedHash.split('$');
      if (parts.length !== 2) return false;

      const salt = parts[0];
      const originalHash = parts[1];

      // Hashear contraseña ingresada con el mismo salt
      const inputHash = this._pbkdf2Like(password, salt);

      // Comparación constante (evita timing attacks)
      return this._constantTimeCompare(inputHash, originalHash);
    } catch (err) {
      console.error('[PasswordService] Error verificando password:', err);
      return false;
    }
  }

  /**
   * Comparación de tiempo constante (evita timing attacks)
   * Siempre tarda el mismo tiempo sin importar dónde falle
   */
  static _constantTimeCompare(a, b) {
    if (a.length !== b.length) return false;

    let result = 0;
    for (let i = 0; i < a.length; i++) {
      result |= a.charCodeAt(i) ^ b.charCodeAt(i);
    }

    return result === 0;
  }

  /**
   * Validar fortaleza de contraseña
   * Retorna: { isValid: boolean, message: string }
   */
  static validatePasswordStrength(password) {
    if (!password || password.length === 0) {
      return { isValid: false, message: 'Contraseña no puede estar vacía' };
    }

    if (password.length < 6) {
      return {
        isValid: false,
        message: 'Contraseña debe tener al menos 6 caracteres'
      };
    }

    const hasUppercase = /[A-Z]/.test(password);
    const hasLowercase = /[a-z]/.test(password);
    const hasNumbers = /[0-9]/.test(password);
    const hasSpecial = /[!@#$%^&*]/.test(password);

    const requirements = [hasUppercase, hasLowercase, hasNumbers, hasSpecial].filter(
      (req) => req
    ).length;

    // Requerimientos mínimos (al menos 3 de 4)
    if (requirements < 3) {
      return {
        isValid: false,
        message:
          'Contraseña debe contener mayúsculas, minúsculas, números y caracteres especiales'
      };
    }

    return { isValid: true, message: 'Contraseña fuerte ✓' };
  }
}

module.exports = PasswordService;

File: mobile_app/lib/services/password_service.dart
import 'dart:convert';
import 'package:crypto/crypto.dart';

/// Servicio de gestión segura de contraseñas
/// - Hash con salt (PBKDF2 simulado usando SHA-256 iterado)
/// - Nunca almacena passwords en plaintext
/// - Válido para online y offline
class PasswordService {
  static final PasswordService _instance = PasswordService._internal();

  factory PasswordService() {
    return _instance;
  }

  PasswordService._internal();

  // Número de iteraciones para el hash (PBKDF2-like)
  static const int HASH_ITERATIONS = 100000;

  /// Generar un salt aleatorio (16 bytes = 32 caracteres hex)
  static String generateSalt() {
    final random = DateTime.now().microsecondsSinceEpoch.toString();
    return sha256.convert(utf8.encode(random)).toString().substring(0, 32);
  }

  /// Hash seguro de contraseña con salt
  /// Retorna: "salt$hashedPassword" para guardar en BD
  String hashPassword(String password) {
    if (password.isEmpty) throw Exception('Contraseña no puede estar vacía');

    final salt = generateSalt();
    final hashed = _pbkdf2Like(password, salt);

    // Formato: salt$hash (para poder verificar después)
    return '$salt\$${hashed}';
  }

  /// PBKDF2 simulado usando SHA-256 iterado
  /// En producción usar librerías especializadas (pointycastle, argon2, etc)
  String _pbkdf2Like(String password, String salt) {
    String hash = password;

    // Iterar SHA-256 N veces con salt
    for (int i = 0; i < HASH_ITERATIONS; i++) {
      hash = sha256
          .convert(utf8.encode('$hash$salt'))
          .toString()
          .substring(0, 64);
    }

    return hash;
  }

  /// Verificar contraseña contra hash almacenado
  /// Retorna: true si contraseña es correcta, false en caso contrario
  bool verifyPassword(String password, String storedHash) {
    if (storedHash.isEmpty || !storedHash.contains('\$')) {
      return false;
    }

    try {
      // Extraer salt y hash del formato "salt$hash"
      final parts = storedHash.split('\$');
      if (parts.length != 2) return false;

      final salt = parts[0];
      final originalHash = parts[1];

      // Hashear la contraseña ingresada con el mismo salt
      final inputHash = _pbkdf2Like(password, salt);

      // Comparar (usando comparison constante para evitar timing attacks)
      return _constantTimeCompare(inputHash, originalHash);
    } catch (e) {
      print('[PasswordService] Error verificando password: $e');
      return false;
    }
  }

  /// Comparación de tiempo constante (evita timing attacks)
  /// Siempre tarda el mismo tiempo sin importar dónde falle
  bool _constantTimeCompare(String a, String b) {
    if (a.length != b.length) return false;

    int result = 0;
    for (int i = 0; i < a.length; i++) {
      result |= a.codeUnitAt(i) ^ b.codeUnitAt(i);
    }

    return result == 0;
  }

  /// Generar contraseña temporal/aleatoria (para testing)
  static String generateRandomPassword() {
    final chars =
        'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789!@#\$%^&*';
    final random = DateTime.now().microsecondsSinceEpoch;

    String password = '';
    for (int i = 0; i < 12; i++) {
      password += chars[(random + i) % chars.length];
    }
    return password;
  }

  /// Validar fortaleza de contraseña
  /// Retorna: (esValida, mensaje)
  (bool, String) validatePasswordStrength(String password) {
    if (password.isEmpty) {
      return (false, 'Contraseña no puede estar vacía');
    }

    if (password.length < 6) {
      return (false, 'Contraseña debe tener al menos 6 caracteres');
    }

    final hasUppercase = password.contains(RegExp(r'[A-Z]'));
    final hasLowercase = password.contains(RegExp(r'[a-z]'));
    final hasNumbers = password.contains(RegExp(r'[0-9]'));
    final hasSpecial = password.contains(RegExp(r'[!@#\$%^&*]'));

    // Requerimientos mínimos (al menos 3 de 4)
    final meetsRequirements =
        [
          hasUppercase,
          hasLowercase,
          hasNumbers,
          hasSpecial,
        ].where((req) => req).length >=
        3;

    if (!meetsRequirements) {
      return (
        false,
        'Contraseña debe contener mayúsculas, minúsculas, números y caracteres especiales',
      );
    }

    return (true, 'Contraseña fuerte ✓');
  }
}
